### ISSUE TITLE
How to use the mesh to generate the mask.

### ISSUE THREAD

**Issue by zshyang** [facebookresearch/meshrcnn#23]
Hi,
I was trying to use my own dataset. So I need to verify the generated training data, like the rotation and translation matrix, are correct.
So I used the `Pix3D` dataset and tried to generate the mask using the ground truth mesh, the rotation matrix, the translation matrix, and the camera matrix. But I could not generate the wanted mask [here](https://github.com/zshyang/meshrcnn-debug/blob/master/2020-03-29%2010-57-15%20%E7%9A%84%E5%B1%8F%E5%B9%95%E6%88%AA%E5%9B%BE.png).
The brief code is like this.
```python
# Define names.
obj_file_name = record["annotations"][0]['mesh']
image_path = record["file_name"]
R = record["annotations"][0]["R"] 
t = record["annotations"][0]["t"] 
K = np.array(
    [[record["annotations"][0]["K"][0],         0,
record["annotations"][0]["K"][1]],
     [        0, record["annotations"][0]["K"][0],
record["annotations"][0]["K"][2]],
     [        0,         0,      1]],
    dtype=np.float32
)

# Load the obj file.
v, f = load_obj(obj_file_name)

# Read the image.
img = imread(image_path)

# Convert the vertices to camera coordinates.
v = convert_vert_to_camera_coords(v, K, R, t)

# Plot the contour on the image.
img = visualize_contour(img, v, f, "obj")

# Visualize the image.
cv2_imshow(img)

def convert_vert_to_camera_coords(v, K, R, t):
    """Convert the vertices in world coordinate to camera coordinates.

    Args:
        v: The vertices array.
        K: The camera matrix.
        R: The rotation matrix. (3 X 3 list)
        t: The translation matrix. (3 list)

    Outputs:
        v: The camera coordinate.
    """

    # Make the vertice upside down.

    # Convert R and t to numpy array.
    R = np.asarray(R)
    t = np.asarray(t)

    # Compute the mesh vertices positions on the image.
    Rt = np.eye(4)
    Rt[:3, 3] = t
    Rt[:3, :3] = R
    Rt = Rt[:3, :] # Convert 4 X 4 to 3 X 4
    P = np.ones((v.shape[0], v.shape[1]+1))
    P[:, :-1] = v
    P = P.T # 4 X N
    img_cor_points = np.dot(K, np.dot(Rt, P))
    img_cor_points = img_cor_points.T # N X 3
    img_cor_points[:, 0] /= img_cor_points[:, 2]
    img_cor_points[:, 1] /= img_cor_points[:, 2]

    # Convert the coordinates to pixel level.
    img_cor_points = img_cor_points.astype(int)

    return img_cor_points
```

**Comment by gkioxari** [comment]
Unfortunately, I can't practically help you very much without access to your data. I think your issue is that you haven't figured out the right transforms from `world space -> camera space -> image space` that would allow you to align the data to the image. 

In general, figuring out these details when working with 3D data is perhaps the most time consuming part of a project! It took me a while to figure it out for Pix3D and I have provided all the code and transforms that will take the raw 3D data from Pix3D and align it to the image or cropped image regions, which is found in the data loaders and data transforms in the codebase. If your data deviates from the Pix3D assumptions/conventions, then you have to do this work from scratch.

**Comment by zshyang** [comment]
After carefully re-read the code. I found that for `Pix3D` the camera coordinate is 180 degree rotated comparing to common definition.
Again, thanks a lot for your help.

**Comment by chicleee** [comment]
> After carefully re-read the code. I found that for `Pix3D` the camera coordinate is 180 degrees rotated comparing to the common definition.
> Again, thanks a lot for your help.


Hi！Dear @zshyang and @gkioxari .
Have you successfully trained Mesh R-CNN on custom dataset?
 
I was trying to train custom dataset too. I visualized the minibatches during training, but there was nothing on 'Voxel' and 'Mesh' subplot. It seems that I didn't figure out the corrent matrixs of my dataset. I'm very confused about it.

Could you help me with the "180 degrees rotated comparing to the common definition" problem ，or share your detailed code about it?

Thanks a million!

**Comment by zshyang** [comment]
Hi @chicleee ,
What I would suggest is understand Mesh R-CNN's projection first. Try to project the provided mesh, rotation matrix, and translation matrix to get the mask then you would probably understand why.

The standard projection formula I think is:


$$
\begin{bmatrix}
x_{img} \\
y_{img} \\
z_{img}
\end{bmatrix}
= 
\begin{bmatrix}
f & 0 & c_{x} \\
0 & f & c_{y}\\
0 & 0 & 1
\end{bmatrix}
\begin{bmatrix}
r_1 & r_2 & r_3 & t_1 \\
r_4 & r_5 & r_6 & t_2\\
r_7 & r_8 & r_9 & t_3
\end{bmatrix}
\begin{bmatrix}
X_{world} \\
Y_{world} \\
Z_{world} \\
1
\end{bmatrix}
$$

The projection formula used in `Pix3D` is:
$$
\begin{bmatrix}
x_{img} \\
y_{img} \\
z_{img}
\end{bmatrix}
= 
\begin{bmatrix}
-f & 0 & c_{x} \\
0 & -f & c_{y}\\
0 & 0 & 1
\end{bmatrix}
\begin{bmatrix}
r_1 & r_2 & r_3 & t_1 \\
r_4 & r_5 & r_6 & t_2\\
r_7 & r_8 & r_9 & t_3
\end{bmatrix}
\begin{bmatrix}
X_{world} \\
Y_{world} \\
Z_{world} \\
1
\end{bmatrix}
$$
